rust/azure_iot_operations_services/src/azure_device_registry.rs

```diff
@@ -0,0 +1,1049 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Types for Azure Device Registry operations.
+
+use core::fmt::Debug;
+use std::collections::HashMap;
+use thiserror::Error;
+
+use crate::common::dispatcher::Receiver;
+use azure_iot_operations_mqtt::interface::AckToken;
+use azure_iot_operations_protocol::common::aio_protocol_error::AIOProtocolError;
+
+use crate::azure_device_registry::adr_name_gen::adr_base_service::client as adr_name_gen;
+    Asset as GenAsset, AssetDataPointObservabilityModeSchema, AssetDataPointSchemaElementSchema,
+    AssetDatasetSchemaElementSchema, AssetEndpointProfile as GenAssetEndpointProfile,
+    AssetEndpointProfileSpecificationSchema as GenAssetEndpointProfileSpecificationSchema,
+    AssetEndpointProfileStatus as GenAssetEndpointProfileStatus,
+    AssetEndpointProfileUpdateEventTelemetry as GenAssetEndpointProfileUpdateEventTelemetry,
+    AssetEventObservabilityModeSchema, AssetEventSchemaElementSchema,
+    AssetSpecificationSchema as GenAssetSpecificationSchema, AssetStatus as GenAssetStatus,
+    AssetUpdateEventTelemetry as GenAssetUpdateEventTelemetry,
+    AuthenticationSchema as GenAuthenticationSchema, DatasetsSchemaSchemaElementSchema,
+    DetectedAsset as GenDetectedAsset, DetectedAssetDataPointSchemaElementSchema,
+    DetectedAssetDatasetSchemaElementSchema, DetectedAssetEventSchemaElementSchema,
+    DetectedAssetResponseStatusSchema, Error as GenError, EventsSchemaSchemaElementSchema,
+    MessageSchemaReference as GenMessageSchemaReference, MethodSchema, RetainSchema,
+    Topic as GenTopic, UsernamePasswordCredentialsSchema as GenUsernamePasswordCredentialsSchema,
+    X509credentialsSchema as GenX509credentialsSchema,
+};
+use crate::azure_device_registry::adr_type_gen::aep_type_service::client as adr_type_gen;
+    DiscoveredAssetEndpointProfile as GenDiscoveredAssetEndpointProfile,
+    DiscoveredAssetEndpointProfileResponseStatusSchema,
+    SupportedAuthenticationMethodsSchemaElementSchema,
+};
+
+/// Azure Device Registry generated code
+mod adr_name_gen;
+mod adr_type_gen;
+/// Azure Device Registry Client implementation wrapper
+mod client;
+
+pub use client::Client;
+
+/// Represents an error that occurred in the Azure Device Registry Client implementation.
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct Error(#[from] ErrorKind);
+
+impl Error {
+    /// Returns the [`ErrorKind`] of the error.
+    #[must_use]
+    pub fn kind(&self) -> &ErrorKind {
+        &self.0
+    }
+}
+
+/// Represents the kinds of errors that occur in the Azure Device Registry Client implementation.
+#[derive(Error, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum ErrorKind {
+    /// An error occurred in the AIO Protocol. See [`AIOProtocolError`] for more information.
+    #[error(transparent)]
+    AIOProtocolError(#[from] AIOProtocolError),
+    /// An error occurred during serialization of a request.
+    #[error("{0}")]
+    SerializationError(String),
```

Comment:
Why is this a string? Why doesn't it convert from the underlying serialization error in protocol using the `#[from]` macro?

Reply:
Needs a bigger discussion - as of now this is the same pattern followed in schema regsitry